packages/firestore/src/local/indexeddb_mutation_queue.ts

```diff
@@ -598,6 +578,58 @@ export class IndexedDbMutationQueue implements MutationQueue {
   }
 }
 
+/**
+ * Delete a mutation batch and the associated document mutations.
+ * @return done, A PersistencePromise that signals when the deletion has completed, and removedDocuments, the document
+ *         mutations that were removed.
+ */
+export function removeMutationBatch(
+  txn: SimpleDbTransaction,
+  userId: string,
+  batch: MutationBatch
+): { removedDocuments: DocumentKey[]; done: PersistencePromise<void> } {
+  const mutationStore = txn.store<DbMutationBatchKey, DbMutationBatch>(
+    DbMutationBatch.store
+  );
+  const indexTxn = txn.store<DbDocumentMutationKey, DbDocumentMutation>(
+    DbDocumentMutation.store
+  );
+  const promises: Array<PersistencePromise<void>> = [];
+
+  const range = IDBKeyRange.only(batch.batchId);
+  let numDeleted = 0;
+  const removePromise = mutationStore.iterate(
+    { range },
+    (key, value, control) => {
+      numDeleted++;
+      return control.delete();
+    }
+  );
+  promises.push(
+    removePromise.next(() => {
+      assert(
+        numDeleted === 1,
+        'Dangling document-mutation reference found: Missing batch ' +
+          batch.batchId
+      );
+    })
+  );
+  const removedDocuments: DocumentKey[] = [];
+  for (const mutation of batch.mutations) {
+    const indexKey = DbDocumentMutation.key(
+      userId,
+      mutation.key.path,
+      batch.batchId
+    );
+    promises.push(indexTxn.delete(indexKey));
+    removedDocuments.push(mutation.key);
+  }
+  return {
+    done: PersistencePromise.waitFor(promises),
+    removedDocuments
+  };
```

Comment:
Isn't this just `return PersistencePromise.waitFor(promises).then(() => removedDocuments)`?

Reply:
Oh, yeah, I like that better. Done.